Scripts: Enable React Fast Refresh for block development

gziolo opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/master/docs/contributors/repository-management.md#pull-requests. -->

## Description
<!-- Please describe what you have changed or added -->

Part of #21008 targeting `@wordpress/scripts`.

I followed instructions at https://github.com/pmmmwh/react-refresh-webpack-plugin and it fully works as expected for Chrome and Safari. It should work with both React Developer Tools enabled but also without them.

I was testing with `SCRIPT_DEBUG` set to `true` for WordPress to ensure that development versions of `react` and `react-dom` were served. In general, the challenge is that we are using externals for both libraries, while in the examples shared they are all bundled together. 


At the moment React Refresh needs to be activated with the `--hot` flag when executing the `start` command from `@wordpress/scripts`:

```bash
wp-scripts start --hot
```

If the flag is not provided then the `start` command should work as before.

## Resolved Blockers

- [x] Hot module replacement for CSS isn't working out of the box with webpack 4 and `MiniCSSExtractPlugin` but it should with webpack 5, see https://webpack.js.org/plugins/mini-css-extract-plugin/#hot-module-reloading-hmr.
- [x] Hot module replacement doesn't work with `MiniCSSExtractPlugin` and `splitChunks` config for the `style.scss` files
- ~Add `react/jsx-runtime` entry for the new automatic JSX transform from React 17 #34221~
- [x] Ensure that React Refresh entry and runtime get externalized and initialized before React loads (https://github.com/pmmmwh/react-refresh-webpack-plugin/blob/main/docs/TROUBLESHOOTING.md#externalising-react).

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

The way I test it:

```bash
npx wp-create-block my-block --no-wp-scripts
cd my-block
../node_modules/.bin/wp-scripts start --hot
```

This is to ensure that we are using the latest `@wordpress/scripts` from the branch and with the default webpack config from `@wordpress/scripts`.

Ensure your WordPress instance has `SCRIPT_DEBUG` set to `true`.

## Screenshots

https://user-images.githubusercontent.com/699132/143570765-0e1f0f25-21c8-4ded-8223-1756103b7895.mov

https://user-images.githubusercontent.com/699132/143570743-1c6a57d3-69e3-44d7-a073-051fc6d0c744.mov



## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
New feature (non-breaking change which adds functionality).
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

gziolo committed:
Scripts: Enable React Fast Refresh for block development

gziolo commented on packages/scripts/config/webpack.config.js:
```diff
@@ -135,7 +136,6 @@ const config = {
 				test: /\.jsx?$/,
 				exclude: /node_modules/,
 				use: [
-					require.resolve( 'thread-loader' ),
```
It might be fine to keep `thread-loader` but I removed it temporarily until the integration works. We can enable it back later.

gziolo replied:
I read at https://webpack.js.org/loaders/thread-loader/:
> Each worker is a separate node.js process, which has an overhead of ~600ms. There is also an overhead of inter-process communication.
>
> Use this loader only for expensive operations!

Let's remove this loader and simplify things.

gziolo commented on packages/scripts/config/webpack.config.js:
```diff
@@ -249,6 +251,15 @@ if ( ! isProduction ) {
+			// Requests come from the WP port.
+			'Access-Control-Allow-Origin': '*',
+		},
+		liveReload: false,
```
`liveReload` override shouldn't be necessary anymore.

gziolo commented:
The current status can be summarized with the following demo:

https://user-images.githubusercontent.com/699132/111445372-a2c2d680-870b-11eb-8cf0-89c6ade74df0.mov

It works perfectly fine for JavaScript files when developing a block 🎉 

There are some issues with how webpack 4 works when `MiniCSSExtractPlugin` is integrated. In particular, it fails miserably with `optimization.splitChunks.cacheGroups` for shared styles.

iandunn commented:
That's awesome, thanks!

I had trouble getting it setup on a site that's served from a local domain w/ SSL (e.g., `https://wp.test`). Webpack's dev server was running on `http://localhost`, but the browser was requesting `https://localhost`. It looks like [overriding the config to use `127.0.0.1` as the hostname](https://github.com/webpack/webpack-dev-server/blob/2ec81605127cec82fae5064dd59da2798a628e02/client-src/default/utils/createSocketUrl.js#L66) might work, but I ran out of time to try that.

gziolo commented:
> I had trouble getting it setup on a site that's served from a local domain w/ SSL (e.g., `https://wp.test`). Webpack's dev server was running on `http://localhost`, but the browser was requesting `https://localhost`. It looks like [overriding the config to use `127.0.0.1` as the hostname](https://github.com/webpack/webpack-dev-server/blob/2ec81605127cec82fae5064dd59da2798a628e02/client-src/default/utils/createSocketUrl.js#L66) might work, but I ran out of time to try that.

I guess this line needs to be improved in the webpack config:

```
publicPath: `http://localhost:${ DEV_SERVER_PORT }/build/`,
```

I'm also not too fond of:
```js
headers: {
 	// Requests come from the WP port.
 	'Access-Control-Allow-Origin': '*',
 },
```

So it's definitely where it would have to improve to support different scenarios available other than using `@wordpress/env` that I used for testing.

iandunn commented:
Ah, yeah, that got me further. It looks like changes are need to both `config.output` and `config.devServer`:

```diff
diff --git a/packages/scripts/config/webpack.config.js b/packages/scripts/config/webpack.config.js
index c33d154..5eb2655 100644
--- a/packages/scripts/config/webpack.config.js
+++ b/packages/scripts/config/webpack.config.js
@@ -8,6 +8,9 @@ const path = require( 'path' );
 const ReactRefreshWebpackPlugin = require( '@pmmmwh/react-refresh-webpack-plugin' );
 const TerserPlugin = require( 'terser-webpack-plugin' );
 
+const fs = require( 'fs' );
+
 /**
  * WordPress dependencies
  */
@@ -89,7 +92,9 @@ const config = {
 		// are used on the same page.
 		// @see https://github.com/WordPress/gutenberg/issues/23607
 		jsonpFunction: getJsonpFunctionIdentifier(),
-		publicPath: `http://localhost:${ DEV_SERVER_PORT }/build/`,
+		publicPath: `https://learn.wordpress.test:${ DEV_SERVER_PORT }/build/`,
 	},
 	resolve: {
 		alias: {
@@ -255,10 +260,20 @@ if ( ! isProduction ) {
 		headers: {
 			// Requests come from the WP port.
 			'Access-Control-Allow-Origin': '*',
 		},
 		liveReload: false,
+		host: 'learn.wordpress.test',
 		port: DEV_SERVER_PORT,
 		writeToDisk: true,
+
+		https: {
+			key: fs.readFileSync( '/Users/iandunn/vhosts/localhost/learn.wordpress.test/learn.wordpress.test-key.pem' ),
+			cert: fs.readFileSync( '/Users/iandunn/vhosts/localhost/learn.wordpress.test/learn.wordpress.test.pem' ),
+			ca: fs.readFileSync( '/Users/iandunn/Library/Application Support/mkcert/rootCA.pem' )
+		}
 	};
 }
 ```

I just hardcoded those for testing, but plugins could set cross-environment things by overriding `webpack.config.js`, and individual devs could set environment-specific things with CLI args.

That gets rid of the JS-related console errors during the initial page load. I still get errors about stylesheets, but that seems to be expected at the moment.

I haven't gotten it working with updates, though.  Each save to `edit.js` gets this in the console:

```
[HMR] Checking for updates on the server... log.js:24
[HMR] Checking for updates on the server... log.js:24
[HMR] Checking for updates on the server... log.js:24
[HMR] Checking for updates on the server... log.js:24
[HMR] Checking for updates on the server... log.js:24
[HMR] Nothing hot updated. log.js:24
[HMR] App is up to date. log.js:24
[HMR] Nothing hot updated. log.js:24
[HMR] App is up to date. log.js:24
[HMR] Nothing hot updated. log.js:24
[HMR] App is up to date. log.js:24
[HMR] Nothing hot updated. log.js:24
[HMR] App is up to date. log.js:24
[HMR] Updated modules: log.js:24
[HMR]  - ./js/workshop/src/edit.js log.js:24
[HMR] App is up to date. log.js:24
```

The watcher sees the edit to `edit.js`, but the changes aren't reflected 🤔

gziolo commented:
> I just hardcoded those for testing, but plugins could set cross-environment things by overriding webpack.config.js, and individual devs could set environment-specific things with CLI args.

It looks like you can set some of those custom settings with CLI, not sure how it would work with `https` though:
https://github.com/webpack/webpack-dev-server#with-the-cli

> I haven't gotten it working with updates, though. Each save to edit.js gets this in the console:

Yes, it might be because of issues with CSS imports. In my testing, I removed all references to all files that match `style.*css` pattern. I hope it will work out of the box with webpack 5.

gziolo commented:
I rebased this branch with the latest changes from `trunk` including webpack migration to v5. 

## Resolves issues

I managed to resolve two known issues that existed before sp CSS reloading works perfectly fine in this branch:

> - Hot module replacement for CSS isn't working out of the box with webpack 4 and MiniCSSExtractPlugin but it should with webpack 5, see https://webpack.js.org/plugins/mini-css-extract-plugin/#hot-module-reloading-hmr.
> - Hot module replacement doesn't work with MiniCSSExtractPlugin and splitChunks config for the style.scss files

## New issues

It looks like hot reloading still works for JavaScript changes, but it doesn't refresh updated React components anymore. At least for my test scenario:

```js
npx wp-create-block gutenpride --no-wp-scripts -t wp-block-directory-template
cd gutenpride
../node_modules/.bin/wp-scripts start --hot
```

The comment with the default webpack config shortcomings is still relevant: https://github.com/WordPress/gutenberg/pull/28273#issuecomment-801670845.

gziolo commented:
I rebased the branch with the latest changes from `trunk` and updated some dependencies.

React refresh works in Chrome, but it still doesn't work in Safari. 

We still might try to externalize React Fast Refresh and put it before `react` as explained in https://github.com/pmmmwh/react-refresh-webpack-plugin/blob/main/docs/TROUBLESHOOTING.md#externalising-react.

gziolo commented on packages/scripts/config/webpack.config.js:
```diff
@@ -265,6 +260,21 @@ if ( ! isProduction ) {
+			writeToDisk: true,
+		},
+		allowedHosts: 'auto',
+		host: 'localhost',
+		port: 8887,
+		proxy: {
+			'/build': {
+				pathRewrite: {
+					'^/build': '',
+				},
+			},
+		},
```
This is the best I could get so far. It automatically works with `ReactRefreshWebpackPlugin` for both Safari and Chrome. The only challenge would be when someone wants to use the port 8887 for something else. I choosed a port number similar to what `@wordpress/wp-env` uses by default: 8888 and 8889.

gziolo commented:
I rebased this PR with `trunk` and improved the dev server configuration. I'm still looking into the issue in Safari which prevents updating React components.

gziolo committed:
Externalize React Refresh to make it work without Dev Tools

gziolo commented:
This PR is ready for review. It works now with both Safari and Chrome without the need to have React Development Tools installed 🎉 


https://user-images.githubusercontent.com/699132/143573548-4b282358-1e95-4682-b39d-95c9123e6fa6.mov


https://user-images.githubusercontent.com/699132/143573569-828f577a-ef6f-4576-adf0-53cdab32e863.mov

At the moment React Refresh needs to be activated with the `--hot` flag when executing the `start` command from `@wordpress/scripts`:

```bash
wp-scripts start --hot
```

If the flag is not provided then the `start` command should work as before.

cbravobernal commented:
Awesome! This will increase our development speed 🎉 
I'm trying to run it but I'm getting a blank page both while I'm trying to [edit a post](http://localhost:8888/wp-admin/post.php?post=1&action=edit) or trying to access through the [URL provided on the console](http://localhost:8081/)

Is there anything I'm missing? I have SCRIPT_DEBUG enabled.

Again, thanks a lot!

Attached screenshots:

<img width="337" alt="Screenshot 2021-11-26 at 14 08 11" src="https://user-images.githubusercontent.com/37012961/143586149-b4ab1d97-dade-4c8a-9d8d-6bc0868c962c.png">

<img width="853" alt="Screenshot 2021-11-26 at 14 08 19" src="https://user-images.githubusercontent.com/37012961/143586165-0bcf45df-0188-429e-902d-a20cf7dfac60.png">

gziolo commented:
> Awesome! This will increase our development speed 🎉 I'm trying to run it but I'm getting a blank page both while I'm trying to [edit a post](http://localhost:8888/wp-admin/post.php?post=1&action=edit) or trying to access through the [URL provided on the console](http://localhost:8081/)

Testing is a bit more complex as it works only for block development or plugin development when using a `@wordpress/scripts` package in 3rd party project. I picked the port 8887 so you should see http://localhost:8887/ instead of http://localhost:8081 on the terminal.
 
> <img alt="Screenshot 2021-11-26 at 14 08 11" width="337" src="https://user-images.githubusercontent.com/37012961/143586149-b4ab1d97-dade-4c8a-9d8d-6bc0868c962c.png">

You can browse files when you know the exact URL. I will provide more details later.

> <img alt="Screenshot 2021-11-26 at 14 08 19" width="853" src="https://user-images.githubusercontent.com/37012961/143586165-0bcf45df-0188-429e-902d-a20cf7dfac60.png">

I have a feeling that you executed a local copy of `wp-scripts` from the root of the Gutenberg project and it uses default setting for the `webpack-dev-server`. You need to make sure that your project uses the default webpack config shipped with `@wordpress/scripts`:

https://github.com/WordPress/gutenberg/blob/dc4f4d8444d87f27095b4f4856d675aa1871c31b/packages/scripts/config/webpack.config.js

I think I need to vastly improve testing instructions because it's a bit complex before we land necessary changes in the Gutenberg plugin and publish `@wordpress/scripts` to npm.

gziolo commented:
I recorded a video in which I explained how this PR could be tested. It is a bit complex to configure because the target group is developers using `@wordpress/scripts` to develop blocks. It is available on YouTube (it was too large to embed it here – 10 minutes):

https://www.youtube.com/watch?v=SbM58ZlwAZM

gziolo commented:
I don’t see any new feedback, so I assume no blockers were discovered. The demo included should be a good confirmation that everything works without issues. New functionality is behind flags so that it won’t impact production builds. There are going to be two new JS files included in the plugin’s zip file, but they are going to be enqueued only when `SCRIPT_DEEBUG` is set to `true`.

michalczaplinski commented:
Awesome work there @gziolo 🙂 

How difficult would it be in your estimation to enable Fast Refresh for development in the Gutenberg repo itself, for example for the blocks in our `block-library` (and not only for the third-party blocks) ?

gziolo commented:
> How difficult would it be in your estimation to enable Fast Refresh for development in the Gutenberg repo itself, for example for the blocks in our `block-library` (and not only for the third-party blocks) ?

Thank you, @michalczaplinski! I have left a comment about that in https://github.com/WordPress/gutenberg/discussions/36142#discussioncomment-1822711. Let me share it here as well:

> It should also be possible to add the same capabilities to the Gutenberg build workflow, but it might require more thinking. We already have a new entry point for React Refresh coming from the webpack plugin and another entry point for React Refresh runtime provided by Facebook. Nest steps would be integrating the webpack dev server with the existing webpack config for the Gutenberg plugin. Another challenge is finding a viable way to include the Babel plugin for React Refresh in the workflow that transpiles WordPress packages.

The webpack config that we use for the Gutenberg plugin comprises 4 subconfigs (one is for WordPress packages, another one for blocks on the frontend), so that might be a concern. I assume that if we make React Fast Refresh work with the `@wordpress/block-library`, then it should also work for all other entry points built from WP packages. The biggest question is how much the workflow will work with parts of the codebase that use an internal state like the registry for blocks. Changing a block definition won't automatically be reflected in JavaScript with hot module reloading. We would have to code our solutions to improve the experience. I guess it's required anyway, and that's the purpose of the Babel plugin for React Fast Refresh.

kevin940726 commented:
For react fast refresh support in Gutenberg repo, it shouldn't be too hard if we have #36736 merged. I'd already gotten that working once in one of my exploration branches. It's outlined in my plans in #36142.

gziolo commented:
`@wordpress/scripts` v20.0.0 is available on npm for downloading.

I did some testing and most of the things work nicely with the block scaffolded with `npx @wordpress/create-block -t @wordpress/create-block-tutorial-template gutenpride`. I use `npm start -- --hot` command. The only issue I see is that changes applied to `src/index.js` trigger full page reload. It looks it's because of any file that doesn't contain any JSX or React code doesn't have hot module reload enabled. I was able to make it work by adding the following code snippet:

```js
import blockMetadata from "../block.json";
if ( module.hot ) {
    module.hot.accept();
    if ( module.hot.status() === 'apply' ) {
        unregisterBlockType( blockMetadata.name );
    }
}
```

It looks like we need some custom Babel code to ensure this case is covered.

Another challenge is that `block.json` is read on the server and propagated to the client. It means that any changes to `block.json` won't get reflected in the editor unless we pass them explicitly when re-registering a block, example:

```js
registerBlockType( blockMetadata.name, {
    ...blockMetadata,
    edit: Edit,
    save,
} );
```

Once we ensure that the metadata is explicitly passed from `block.json` then everything should work just fine. It's another thing that Babel transform could help with.

vena commented:
I'm not too well versed in webpack, but has _not_ using this been overlooked?

The included webpack config seems to only check if the env is production, and if not loads react refresh, omitting the --hot flag in "wp-scripts start" doesn't affect this.  npm warns that HMR isn't loaded, wp-react-refresh-runtime has been added to asset requirements, block startup in the editor now fails with:

`Uncaught TypeError: Cannot read properties of undefined (reading 'injectIntoGlobalHook')
    at Object../node_modules/@pmmmwh/react-refresh-webpack-plugin/client/ReactRefreshEntry.js`

gziolo commented:
@vena, thank you for the report. Can you share exact steps so I could reproduce it? It’s possible that I missed something obvious. I must admit that this integration was very complex and I definitely don’t consider myself webpack expert 😃

vena commented:
I've created a little test plugin project at https://github.com/vena/wps-test - it uses @wordpress/scripts 20.0.0, but all it does is enqueue an editor script that should pop alert('hello!')

first thing that happens when I run npm start (note, no --hot flag in package.json), I get a warning:

`<w> [ReactRefreshPlugin] Hot Module Replacement (HMR) is not enabled! React Refresh requires HMR to function properly.`

Activating the plugin in a fresh WP 5.9 docker container and entering the editor, the first thing I noticed was that the test plugin's enqueued script wasn't loading. For some reason WP won't load it if 'wp-react-refresh-runtime' is in the deps of the wp_enqueue_script call. 

That's weird and I don't know why it's doing that, it wasn't doing that in another plugin i'm working on that made me post my last comment... but I'm not using that dependency anyway, and I get the same error, so I leave it out of the deps array for purposes of demonstration.

Reload the post editor, and dev console reports:

![image](https://user-images.githubusercontent.com/867733/151461211-080efe44-c44d-4e46-8639-66e7a706204c.png)

My alert never gets a chance to run.

jakeparis commented:
Hi @gziolo - thanks so much for working on this, I'm really excited to get it working

* I've got the gutenberg plugin running
* I've enabled `SCRIPT_DEBUG` in wp-config
* I have React Developer Tools browser plugin running (even though I noted you said it shouldn't be necessary).

I develop on my local machine using domains like **https://mywebsite.local**. I have a full instance of wordpress running, and I've working within the block editor.

Currently I'm getting the following error:
```
Firefox can’t establish a connection to the server at wss://localhost:8887/ws.
```
and in chromish:
```
WebSocketClient.js:16 WebSocket connection to 'wss://localhost:8887/ws' failed: 
```


I've added the following `webpack.config.js` file to the root of my plugin, but it wasn't giving errors before this as well.
```js
const defaultConfig = require( '@wordpress/scripts/config/webpack.config' );

module.exports = {
	...defaultConfig,

	devServer: {
		...defaultConfig.devServer,

		https: true,
		allowedHosts: 'all', // also tried setting to "mywebsite.local"
		headers: {
			"Access-Control-Allow-Origin": "*",
			"Access-Control-Allow-Credentials": "true",
			"Access-Control-Allow-Methods": "GET, POST, PUT, DELETE, OPTIONS",
		},
	},

};
```

Do you have any insight on what I might have setup wrong?

jakeparis commented:
I also tried adding `host: "mywebsite.local"` to the devServer config, but no luck there either.

jakeparis commented:
Oops, solved it myself by just shooting in the dark. 
Fixed with this webpack.config.js:
```js
const defaultConfig = require( '@wordpress/scripts/config/webpack.config' );

module.exports = {
	...defaultConfig,

	devServer: {
		...defaultConfig.devServer,

		allowedHosts: 'all',

	},

};
```

I do still get the error message about not being able to connect, but the hot refresh is working.

gziolo commented:
@jakeparis, you solved it. That's awesome. 🎉 

I tested with http://localhost:8888 and http://localhost:8889. It looks like the default setup is too restrictive, but I think it wasn't recommended to use `allowedHosts: 'all'` because of security concerns 😅 

You can also pass regular CLI flags to `wp-scripts start` that webpack dev server support to override without touching the config file. I need to research it further but it looks like we should either document your finding or change allowed hosts. See https://webpack.js.org/configuration/dev-server/#devserverallowedhosts.

dgwyer commented:
@gziolo Yes, please document how to get this to work for other hosts. :) That wold be most helpful.

I'm personally using DesktopServer and couldn't get fast refresh to work. I assumed it was some sort of host proxy that needed adding. My local host URL is typically something like: https://www.example.dev.cc/.

I also use LocalWP by Flywheel but not tested with that yet.

gziolo commented:
@dgwyer, I must admit that juggling with those host-related settings took me a ton of time. WordPress isn't a regular target for weback dev server where you usually server everything bundled together into a single app 😄 

There is also room for improvement for how hot module reloading works with blocks as documented in https://github.com/WordPress/gutenberg/pull/28273#issuecomment-1023340680. This is why it still needs to be explicitly activated. I hope we all sort it out together.

jakeparis commented:
Actually, adding the devServer config in webpack.config.js broke my build pipline. Doing is this way solved that issue:
```js

const defaultConfig = require( '@wordpress/scripts/config/webpack.config' );

if( defaultConfig.mode === 'development' ) 
	defaultConfig.devServer.allowedHosts = "all";

module.exports = defaultConfig;
```

gziolo commented:
From https://webpack.js.org/configuration/dev-server/#devserverallowedhosts:

> When set to 'auto' this option always allows local[host](https://webpack.js.org/configuration/dev-server/#devserverhost), host, and [client.webSocketURL.hostname](https://webpack.js.org/configuration/dev-server/#websocketurl):

Maybe you could run it as `wp-scripts start --hot --host mywebsite.local --port auto` without overriding the webpack config. I'm guessing only based on the current configuration:

https://github.com/WordPress/gutenberg/blob/9482dfa2f9d9820fa4177e4f4ece6c4c4c391b79/packages/scripts/config/webpack.config.js#L274-L275

jakeparis commented:
@vena, 
I got this error when I didn't have the gutenberg plugin activated. Are you sure you have the gutenberg plugin activated? 

> I've created a little test plugin project at https://github.com/vena/wps-test - it uses @wordpress/scripts 20.0.0, but all it does is enqueue an editor script that should pop alert('hello!')
> 
> first thing that happens when I run npm start (note, no --hot flag in package.json), I get a warning:
> 
> `<w> [ReactRefreshPlugin] Hot Module Replacement (HMR) is not enabled! React Refresh requires HMR to function properly.`
> 
> Activating the plugin in a fresh WP 5.9 docker container and entering the editor, the first thing I noticed was that the test plugin's enqueued script wasn't loading. For some reason WP won't load it if 'wp-react-refresh-runtime' is in the deps of the wp_enqueue_script call.
>

vena commented:
> @vena, I got this error when I didn't have the gutenberg plugin activated. Are you sure you have the gutenberg plugin activated?

@jakeparis  I did not have the gutenberg plugin active at any point. I wasn't using or intending to use HMR, so I omitted --hot from the start script as per the README and shouldn't need the gutenberg plugin, right?

shivapoudel commented:
@gziolo On Local by Flywheel this is not working with the custom domain in HTTPS mode. How can I achieve this feature?

gziolo commented:
> On Local by Flywheel this is not working with the custom domain in HTTPS mode. How can I achieve this feature?

I would try a similar trick like in https://github.com/WordPress/gutenberg/pull/28273#issuecomment-1036511224. You probably will have to override more settings to ensure that the dev server and your WP instance can talk to each other. It would be great to see people sharing their success stories so we can figure out something more general. It's one of the trickiest parts here.

shivapoudel commented:
@gziolo it works with that. I am currently using this:

```
"start:hot": "wp-scripts start --hot --allowed-hosts all",
```

nicholasio commented:
@gziolo Thanks for all the great work. I've been working to implement this on 10up's fork of wp-scripts (https://github.com/10up/10up-toolkit/pull/150) and noticed something:

It seems like compiling `react-refresh-entry` is not actually needed https://github.com/WordPress/gutenberg/blob/trunk/tools/webpack/development.js#L24 and I'm not sure if you're even loading this in WP at all?

gziolo commented:
@nicholasio, I'm very excited to see it being implemented at 10up 🎉 I'm happy to help you to make it work so we could learn from each other.

> It seems like compiling `react-refresh-entry` is not actually needed https://github.com/WordPress/gutenberg/blob/trunk/tools/webpack/development.js#L24 and I'm not sure if you're even loading this in WP at all?

It depends on the browser used. It isn't necessary in Chome when React Developer Tools are enabled. I see a comment from @fabiankaegy in the PR referenced above that there are some issues in Safari. I left the following comment there:

> There are no React Developer Tools for Safari so it's harder to configure it. React Refresh entry and runtime needs to be externalized and initialized before React loads (https://github.com/pmmmwh/react-refresh-webpack-plugin/blob/main/docs/TROUBLESHOOTING.md#externalising-react). At the moment it's only available with the Gutenberg plugin enabled and `SCRIPTS_DEBUG` set to `true` that ensures that the React Refresh runtime is correctly configured. In Chrome, it should work with the Gutenberg plugin enabled and it probably doesn't need the React Refresh runtime.

nicholasio commented:
@gziolo Thanks for following up. re: `react-refresh-entry` yeah we noticed it wasn't working for @fabiankaegy I guess bc he didn't have the react developer tools installed. I have included it again and it's now working.

dgwyer commented:
> @gziolo it works with that. I am currently using this:
> 
> ```
> "start:hot": "wp-scripts start --hot --allowed-hosts all",
> ```

@shivapoudel I'm also trying to get this working with LocalWP but without much success. No matter what settings I use it doesn't hot reload. Is this still working for you? I'm on Windows and am beginning to wonder if this is a factor too.

shivapoudel commented:
@dgwyer You have to activate the Gutenberg plugin to get this working, my friend :)

gziolo commented:
> You have to activate the Gutenberg plugin to get this working, my friend :)

It reminds me that we should bring the same scripts to WordPress core in the 6.0 release cycle. We still will need to put it behind the `SCRIPT_DEBUG` constant.

gziolo commented:
> You have to activate the Gutenberg plugin to get this working, my friend :)

It works now also with WordPress 6.0 with the `SCRIPT_DEBUG` flag enabled.

rafaucau commented:
@gziolo I set `allHosts` and refresh works, but I get this errors in the console:
![image](https://user-images.githubusercontent.com/25438601/183757820-be3e034b-76cd-4ec2-a75d-a07861f1f62a.png)
![image](https://user-images.githubusercontent.com/25438601/183757812-2be14b23-dd9a-4422-9d36-0b973f5a8320.png)

When I update the code (React) of my block, the page refreshes instead of applying the update.

**Edit:**
The `--https` flag was missing `wp-scripts start --hot --allowed-hosts localhost --allowed-hosts .localhost --https`, but still the page refreshes instead of updating live.
![image](https://user-images.githubusercontent.com/25438601/183762259-6cf943a1-eb29-4bdf-bedc-d14159c1c481.png)

gziolo commented:
> The `--https` flag was missing `wp-scripts start --hot --allowed-hosts localhost --allowed-hosts .localhost --https`, but still the page refreshes instead of updating live.

Setting up the connection between the WordPress instance and the dev server is the worst. Unfortunately, I don't have enough experience with that part 😞

rafaucau commented:
I found a solution. I'm using Docker and the site in the localhost subdomain works in the browser, but there was no this subdomain in the `hosts` file. I added it and it started working.

eduwass commented:
@shivapoudel @dgwyer I know I'm late to the party but I can't get this to work in LocalWP, tried all of the solutions in the post I think.


Currently my script has this:
```json
		"start:hot": "wp-scripts start --hot --host officehours.local --allowed-hosts all"
```

It looks like it starts:
```
officehours on  main [!] is 📦 v0.1.0 via ⬢ v19.0.1 via 🐘 v8.1.12 took 1m 33s 
➜ npm run start:hot

> officehours@0.1.0 start:hot
> wp-scripts start --hot --host officehours.local --allowed-hosts all

<i> [webpack-dev-server] Project is running at:
<i> [webpack-dev-server] Server: http://officehours.local:8080/
<i> [webpack-dev-server] Loopback: http://[::1]:8080/
<i> [webpack-dev-server] Content not from webpack is served from '/Users/eduwass/Sites/officehours/app/public/wp-content/plugins/officehours/public' directory
assets by chunk 123 KiB (name: index)
  asset index.js 122 KiB [emitted] [minimized] (name: index)
  asset index.asset.php 139 bytes [emitted] (name: index)
  asset index.css 57 bytes [emitted] (name: index)
asset block.json 491 bytes [emitted] [from: src/block.json] [copied]
asset render.php 142 bytes [emitted] [from: src/render.php] [copied]
...
```

But then I see nothing on the Server URL?

```
officehours on  main [!] is 📦 v0.1.0 via ⬢ v19.0.1 via 🐘 v7.4.30
➜ curl -I http://officehours.local:8080/
HTTP/1.1 404 Not Found
X-Powered-By: Express
Content-Security-Policy: default-src 'none'
X-Content-Type-Options: nosniff
Content-Type: text/html; charset=utf-8
Content-Length: 140
Vary: Accept-Encoding
Date: Thu, 16 Feb 2023 03:46:02 GMT
Connection: keep-alive
Keep-Alive: timeout=5


officehours on  main [!] is 📦 v0.1.0 via ⬢ v19.0.1 via 🐘 v7.4.30
➜ curl http://officehours.local:8080/
<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="utf-8">
<title>Error</title>
</head>
<body>
<pre>Cannot GET /</pre>
</body>
</html>
```

Apologies if I'm missing something obvious! Would love to see this working 😅

dgwyer commented:
I didn't get this working in LocalWP either. Haven't tried for a while though.

@rafaucau

> I found a solution. I'm using Docker and the site in the localhost subdomain works in the browser, but there was no this subdomain in the hosts file. I added it and it started working.

Did you mean you got it to work with wp-env (which is Docker based), or your own Docker solution? Would love to learn more about this.

eduwass commented:
@gziolo any idea on https://github.com/WordPress/gutenberg/pull/28273#issuecomment-1432466728 and if not, do you think it's worth creating a separate issue? thanks!

rafaucau commented:
> @dgwyer  Did you mean you got it to work with wp-env (which is Docker based), or your own Docker solution? Would love to learn more about this.

My own Docker solution based on this: https://github.com/litespeedtech/ols-docker-env
Currently this doesn't work for me if the plugin has multiple Gutenberg blocks. When the plugin has one block (block.json directly in the src folder), it works.

So, it doesn't work:
-src
--block1
---block.json
--block2
---block.json

But it works:
-src
--block.json


With wp-env, both versions works ¯\_(ツ)_/¯

rafaucau commented:
I am getting this error: 
```
[HMR] Update failed: ChunkLoadError: Loading hot update chunk *******/index failed.
(missing: http://localhost/plugins/*******/build/*******/index.735c06a3725cd44a7493.hot-update.js)
    at http://localhost/plugins/*******/build/*******/index.js?ver=d0c04516460110d7e803:11201:26
    at new Promise (<anonymous>)
    at loadUpdateChunk (http://localhost/plugins/*******/build/*******/index.js?ver=d0c04516460110d7e803:11196:20)
    at http://localhost/plugins/*******/build/*******/index.js?ver=d0c04516460110d7e803:11656:29
    at Array.forEach (<anonymous>)
    at __webpack_require__.hmrC.jsonp (http://localhost/plugins/*******/build/*******/index.js?ver=d0c04516460110d7e803:11651:22)
    at http://localhost/plugins/*******/build/*******/index.js?ver=d0c04516460110d7e803:10874:47
    at Array.reduce (<anonymous>)
    at http://localhost/plugins/*******/build/*******/index.js?ver=d0c04516460110d7e803:10870:55
```

But when I click on the address of that chunk, I can see the file in the browser. So there is probably a race condition here. The notification of the change is sent to the browser before Docker has had a chance to copy that file.
Of course, I have the project set up on the WSL file system, not on Windows.

rafaucau commented:
I finally found a solution (at least for me)! The problem is caused by multiple entry points (I have multiple blocks in one plugin). You need to add the following code to `webpack.config.js`:
```js
optimization: {
  runtimeChunk: 'single',
},
```

For example:
```js
const defaultConfig = require('@wordpress/scripts/config/webpack.config');
const { getWebpackEntryPoints } = require('@wordpress/scripts/utils/config');

const isProduction = process.env.NODE_ENV === 'production';

const config = {
  ...defaultConfig,
  entry: {
    ...getWebpackEntryPoints(),
    index: './src/index.ts',
  },
};

if (!isProduction) {
  // Fix HMR
  config.optimization = {
    ...defaultConfig.optimization,
    runtimeChunk: 'single',
  };
}

module.exports = config;

```

This creates an additional `runtime.js` file. You need to load it in the editor:
```php
add_action( 'enqueue_block_assets', function () {
  $runtime_asset_path = plugin_dir_path( __FILE__ ) . 'build/runtime.asset.php';

  if ( file_exists( $runtime_asset_path ) ) {
    $runtime_asset = require $runtime_asset_path;

    wp_enqueue_script(
      'my-plugin-runtime',
      plugin_dir_url( __FILE__ ) . 'build/runtime.js',
      $runtime_asset['dependencies'],
      $runtime_asset['version'],
    );
  }
} );
```

dgwyer commented:
Nice work, thanks for sharing.

gziolo commented:
> I finally found a solution (at least for me)! The problem is caused by multiple entry points (I have multiple blocks in one plugin). You need to add the following code to `webpack.config.js`:
> 
> ```js
> optimization: {
>   runtimeChunk: 'single',
> },
> ```

It took me some time to find my old comment about using runtime chunk added in a different place: https://github.com/WordPress/gutenberg/issues/25188#issuecomment-1104898123. Thank you for sharing your findings. It looks like it still works. It would be great to have this runtime always present in the development mode when using hot module replacement. The only challenge would be to integrate the script enqueuing somehow with `register_block_type` call.

> @gziolo any idea on [#28273 (comment)](https://github.com/WordPress/gutenberg/pull/28273#issuecomment-1432466728) and if not, do you think it's worth creating a separate issue? thanks!

I don't think I can help much here. Have you tried accessing the`build` subfolder on the webpack dev server? It's where files are exposed if I recall correctly. See settings:

https://github.com/WordPress/gutenberg/blob/f64379314802d2d4b64077ca6b980e9df824b762/packages/scripts/config/webpack.config.js#L312L325

eduwass commented:
Hi @gziolo been playing with this a bit today:

> Have you tried accessing the`build` subfolder on the webpack dev server? It's where files are exposed if I recall correctly. See settings:

I created a new site in LocalWP, called **tests1.local**

I have an example plugin running inside it:

<img width="935" alt="CleanShot 2023-03-05 at 18 46 14@2x" src="https://user-images.githubusercontent.com/1070495/222976977-2a227a06-cc19-46e2-9c9d-a63e4e6c9883.png">

As you can see I've added a `webpack.config.js` so I'm able to mess around with config details.

### Starting script

I've added the following script to `package.json`:
```
		"start:hot": "wp-scripts start --webpack-copy-php --hot"
```

When I start it, this is the output

```bash
wp-content/plugins/example-dynamic is 📦 v0.1.0 via ⬢ v19.7.0 via 🐘 v8.1.9
➜ npm run start:hot

> example-dynamic@0.1.0 start:hot
> wp-scripts start --webpack-copy-php --hot

<i> [webpack-dev-server] Project is running at:
<i> [webpack-dev-server] Server: http://tests1.local:8887/
<i> [webpack-dev-server] Loopback: http://[::1]:8887/
<i> [webpack-dev-server] Content not from webpack is served from '/Users/eduwass/Sites/tests1/app/public/wp-content/plugins/example-dynamic/public' directory
assets by chunk 519 KiB (name: index)
  asset index.js 518 KiB [emitted] (name: index) 1 related asset
  asset index.css 959 bytes [emitted] (name: index) 1 related asset
  asset index.asset.php 167 bytes [emitted] (name: index)
asset ./style-index.css 1 KiB [emitted] (name: ./style-index) (id hint: style) 1 related asset
asset block.json 503 bytes [emitted] [from: src/block.json] [copied]
asset render.php 149 bytes [emitted] [from: src/render.php] [copied]
Entrypoint index 520 KiB (537 KiB) = ./style-index.css 1 KiB index.css 959 bytes index.js 518 KiB index.asset.php 167 bytes 3 auxiliary assets
```

### Build folder
I am able to access files from the `/build` folder, for example:

```
➜ curl tests1.local:8887/block.json
{
  "$schema": "https://schemas.wp.org/trunk/block.json",
  "apiVersion": 2,
  "name": "create-block/example-dynamic",
  "version": "0.1.0",
  "title": "Example Dynamic",
  "category": "widgets",
  "icon": "smiley",
  "description": "Example block scaffolded with Create Block tool.",
  "supports": {
    "html": false
  },
  "textdomain": "example-dynamic",
  "editorScript": "file:./index.js",
  "editorStyle": "file:./index.css",
  "style": "file:./style-index.css",
  "render": "file:./render.php"
}%
```

So it seems like webpack is connect properly? Also this:

> [webpack-dev-server] Content not from webpack is served from '/Users/eduwass/Sites/tests1/app/public/wp-content/plugins/example-dynamic/public' directory

...is also true, if I create a `/public/index.html` inside of my plugin folder, I can see it when I do:

```
wp-content/plugins/example-dynamic is 📦 v0.1.0 via ⬢ v19.7.0 via 🐘 v8.1.9
➜ curl tests1.local:8887
<h1>Hello world</h1>
```

It's like webpack is there serving the correct files, but the connection with the browser is not working?

jlandrum commented:
I'm not quite sure how to approach this, an issue may be needed, but I'm not too sure what exactly the issue is and thus how to indicate what needs to be fixed exactly. I've been trying to get fast refresh working, maybe it's because the package itself is unstable (as indicated by the authors) but...

In a project with no changes and multiple blocks:
* Only one block will fast refresh; this seems to be expected due to multiple hot reload runtimes being instantiated.
* Only that same block's CSS updates
* viewScript works (and also does seem to hot reload fine regardless of number of blocks)

After making changes as suggested by @rafaucau :
* Blocks fast refresh just fine
* CSS updates just fine, but only if style is updated to: ["file:./style-index.css", "file:./index.css"]. I'm not even sure why style.scss is becoming style-index.css, tried looking through source but can't find anything indicating why this name change is happening
* viewScript no longer works. Manually editing the generated js and adding a console log shows it's being loaded just fine, but the contents of the script aren't being executed. There's nothing obvious showing why.

I'd like to get this working properly, and contribute back to these scripts as they are extremely useful - any direction would be greatly appreciated so I can open a new issue with clearer details and possibly attempt at fixing this.

Edit:
In case anyone else is still having issues as well, for scope - I am building a tool for rapid theme development where blocks are installed in the theme (with plans to eventually allow this tool to target plugins instead if the user wishes - but the focus is on themes with theme-specific blocks), I realized that in most cases it's only necessary to live reload a single block type at a time, so here's what I'm using:

`wp-scripts start --hot --webpack-src-dir=blocks/${name} --output-path=${getThemeDestination()}/blocks/${name}`

So far it has solved all of my issues without the need to modify anything.

gziolo commented:
@jlandrum, thank you so much for sharing your use case and the way you approached moving forward by building only a single block at the time. It's a known limitation that we would like to resolve finally. I recently opened a pike PR against WordPress Core seeking a way to introduce shared scripts: https://github.com/WordPress/wordpress-develop/pull/5118. So there are two ideas discussed currently:

- We enable support for the `runtime` chunk in the development mode, which ensures that webpack can share its mechanism for controlling chunks between blocks.
- Alternatively, we bundle all blocks together when targeting the editor, which means we remove the same issue using only a single entry point.

@jsnajdr might have some more thoughts to add on how we can address this recurring issue. Once we have it sorted out, we can also explore how to improve hot module reloading for the code that is outside React components. In particular, `registerBlockType` has this limitation in that it can be registered only once with a given name, so when reloading the file, we would have to unregister the block or lift the limitation in that mode. I documented my previous exploration around that part in https://github.com/WordPress/gutenberg/pull/28273#issuecomment-1023340680.

> CSS updates just fine, but only if style is updated to: ["file:./style-index.css", "file:./index.css"]. I'm not even sure why style.scss is becoming style-index.css, tried looking through source but can't find anything indicating why this name change is happening

This is coming from this "optimization": 
https://github.com/WordPress/gutenberg/blob/f35837b391e15d4c95665a7d4b0b87c45f139fb7/packages/scripts/config/webpack.config.js#L129-L142

If I remember correctly, the idea was that there is a single style file extracted for every entry point so the names don't conflict if the JS files end up in the same folder. However, this might not make sense anymore, and it could benefit from an update. The initial reasoning seems to be captured in https://github.com/WordPress/gutenberg/tree/trunk/packages/scripts#using-css. It feels like enforcing a separate folder for every entry point would allow us having something like `["file:./index/style.css"]`, but it's also something that @jsnajdr has recently pointed out as it could get auto-generated in the `block.json` by the `wp-scripts start` / `wp-scripts build` so the path wouldn't matter that much anymore.

gziolo commented:
> Currently this doesn't work for me if the plugin has multiple Gutenberg blocks. When the plugin has one block (block.json directly in the src folder), it works.

Regarding the issue reported in https://github.com/WordPress/gutenberg/pull/28273#issuecomment-1434552486, and the workaround  shared in https://github.com/WordPress/gutenberg/pull/28273#issuecomment-1445471529 by @rafaucau. I have a working PR that resolves it:
- https://github.com/WordPress/gutenberg/pull/64924
